# MATH 484 Linear Programs and Related Problems

MATH 484 Linear Programs and Related Problems

The Pennsylvania State University

Based on Prof. Christopher Byrne's lecture materials and his _Linear Programs I: A Clear, Concise, and Practical Introduction to Linear Programs and Duality Theory with Applications to Return On Investment (ROI)_.

---

## Imports & Environment

In [1]:
import numpy                    as np
import pandas                   as pd

from   datetime import datetime as d
import locale                   as l
import platform                 as p
import sys                      as s

pad = 20
print(f"{'Executed'.upper():<{pad}}: {d.now()}")
print()
print(f"{'Platform'        :<{pad}}: "
      f"{p.mac_ver()[0]} | "
      f"{p.system()} | "
      f"{p.release()} | "
      f"{p.machine()}")
print(f"{''                :<{pad}}: {l.getpreferredencoding()}")
print()
print(f"{'Python'          :<{pad}}: {s.version}")
print(f"{''                :<{pad}}: {s.version_info}")
print(f"{''                :<{pad}}: {p.python_implementation()}")
print()
print(f"{'NumPy'           :<{pad}}: {np.__version__}")
print(f"{'Pandas'          :<{pad}}: {pd.__version__}")

EXECUTED            : 2023-01-18 09:12:50.281421

Platform            : 13.1 | Darwin | 22.2.0 | arm64
                    : UTF-8

Python              : 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:25:13) [Clang 14.0.6 ]
                    : sys.version_info(major=3, minor=10, micro=8, releaselevel='final', serial=0)
                    : CPython

NumPy               : 1.24.0
Pandas              : 1.5.2


---

\begin{align}
\text{max} \,\,\, 20x_1 + 10x_2 & \,\,\, \text{s.t.} \\
30x_1 + 10x_2 &\le 400 \\
 3x_1 +  4x_2 &\le  80 \\
          x_1 &\le  10 \\
          x_2 &\le  40 \\
     x_1, x_2 &\ge   0 \\
\end{align}

---

The goal of an LP is to find the decision that optimizes some objective subject to certain constraints.

---

1.2 Def - LINEAR FUNCTION

A function $f : \mathbb{R}^n \rightarrow \mathbb{R}$ is linear if $f(ax + y) = af(x) + f(y) \,\,\, \forall{x, y} \in \mathbb{R}^n, \forall{a} \in \mathbb{R}$.

Note that if $f : \mathbb{R}^n \rightarrow \mathbb{R}$ is linear, then $f(0) = 0$.

1.3 Def - AFFINE FUNCTION

A function $g : \mathbb{R}$ is affine if $f(x) = g(x) + c \,\,\, \forall{x} \in \mathbb{R}^n$ where $g$ is linear and $c \in \mathbb{R}$.

In other words, an affine function is just a linear function plus a constant.

1.4 Def - LINEAR CONSTRAINT

A linear constraint uses a linear function $f : \mathbb{R}^n \rightarrow \mathbb{R}$ and a constant $b \in \mathbb{R}$ to limit the values of a variable $x \in \mathbb{R}^n$ by requiring either $f(x) \le b$, $f(x) \ge b$, or $f(x) = b$.

Only weak inequalities are allowed, which makes the solution set closed.

Note that any affine constraint can be rewritten as a linear constraint.

1.5 Def - [LP] LINEAR PROGRAM

An LP is a problem of optimizing an affine function subject finitely many linear constraints.

The function to be optimized is called the objective function.

Note that any affine objective function $f(x) + c$ attains its max (or min) where $f(x)$ attains its max (or min).

The constant $c$ only changes the optimal value of $f$, not the values of $x$ at which $f$ attains its maximum (or minimum).

1.6 Def - FEASIBLE SET

The feasible set of an LP is the set of all $x \in \mathbb{R}^n$ that satisfy all the constraints of the LP.

The feasible set may also be called the constraint set or the feasible region.

1.7 Def

$\argmax_{\Gamma}(f) = \{ x \in \Gamma \,\,\, | \,\,\, f(x) \ge f(y) \,\,\, \forall{y} \in \Gamma \}$

$\argmax_{\Gamma}(f)$ is the set of points at which $f$ attains its maximum over the set $\Gamma$.

$\argmin_{\Gamma}(f) = \{ x \in \Gamma \,\,\, | \,\,\, f(x) \le f(y) \,\,\, \forall{y} \in \Gamma \}$

$\argmin_{\Gamma}(f)$ is the set of points at which $f$ attains its minimum over the set $\Gamma$.

1.8 Def - OPTIMAL/SOLUTION SET

If $\Gamma$ is the feasible set of an LP, then the optimal/solution set of the LP is $\argmax_{\Gamma}(f)$, if the LP is to maximize $f$, or $\argmin_{\Gamma}(f)$, if the LP is to minimize $f$.

1.9 Def - OPTIMAL VALUE

The optimal value of a maximization LP is $\text{max}_{\Gamma}(f)$ where $\Gamma$ is the feasible region of the LP.

The optimal value of a minimization LP is $\text{min}_{\Gamma}(f)$ where $\Gamma$ is the feasible region of the LP.

1.10 Def - LP SOLUTION

The solution to an LP consists of the optimal set and the optimal value.

In other words, solving an LP means finding both the optimal set and the optimal value.

1.11 Def - GRADIENT

Let $f : \mathbb{R}^n \rightarrow \mathbb{R}$. The gradient of $f$ is $\nabla{f} = \left( \frac{\partial f}{\partial x_1}, ..., \frac{\partial f}{\partial x_n} \right)$.

Note that if $f$ is linear or affine, then $\nabla{f}$ is constant.

The significance of the vector $\nabla{f}$ is that it points in the direction of the maximum rate of increase of $f$ per unit Euclidean distance change in the variable $x$ as it varies through $\mathbb{R}^n$.

1.12 Def - DIRECTIONAL DERIVATIVE

If $u \in \mathbb{R}^n$ is a nonzero vector and $f : \mathbb{R}^n \rightarrow \mathbb{R}$ is a function, then the directional derivative of $f$ in the direction of $u$ equals $\frac{\nabla{f} \cdot u}{||u||}$.

Note that the sign of the directional derivative does not depend on the magnitude of $u$. So $\nabla{f} \cdot u$ can be used in place of the true directional derivative when only the sign is needed. The sign is enough to tell whether $f$ increases $(\nabla{f} \cdot u \ge 0)$; decreases $(\nabla{f} \cdot u \le 0)$; or remains constant $(\nabla{f} \cdot u = 0)$ when the variable $x$ is changed in the direction of $u$.

Note also that because the gradient of an affine function is constant, the directional derivative in any direction is constant. (This will be relevant for step 2C of the graphical method.)

1.13 Def - LEVEL SET ?????

If $f : \mathbb{R}^n \rightarrow \mathbb{R}$ a level set of $f$ is a subset of $\mathbb{R}^n$ on which $f$ has constant value $f(x) = b$ for some constant $b \in \mathbb{R}$, then $f(x) = b$ will sometimes be used as an abbreviation to refer to the level set $\{ x \in \mathbb{R}^n \,\,\, | \,\,\, f(x) = b \}$. (Whether the reference is to the equation itself or to the solution of the equation should be clear from the context.)

1.14 - ?????

If $f(x) = b$ on a set $S \in \mathbb{R}^n$ and $x, y \in S$ where $u = x - y$, then $\nabla{f} \cdot u = 0$.

That is, a function has zero directional derivative on any level set.

1.14.5 - using gradients to graph lines

The boundary of a linear inequality constraint in $\mathbb{R}^n$ is a hyperplane of dimension $n - 1$.
* In $\mathbb{R}^2$, a hyperplane of dimension $2 - 1 = 1$ is a line.
* In $\mathbb{R}^3$, a hyperplane of dimension $3 - 1 = 2$ is a plane.

The full solution of a linear equality constraint in $\mathbb{R}^n$ is a hyperplane of dimension $n - 1$.

A quick way to graph lines and an easy way to visually check one's graphs of lines in $\mathbb{R}^2$ is that the boundary of a linear constraint $f(x) \le b$ will be the line $f(x) = b$, a line perpendicular to $\nabla{f}$. Furthermore, $f(0) = 0$, so $\nabla{f}$ can be traced from the origin out to the level set $f(x) = b$.

For example,
* the boundary of the constraint $2x_1 + 3x_2 \le 6$ will be perpendicular to $\nabla{f} = (2, 3)$ and pass through the first quadrant with negative slope
* the boundary of the constraint $2x_1 - 3x_2 \le 6$ will be perpendicular to $\nabla{f} = (2, -3)$ and pass through the fourth quadrant with positive slope
* the boundary of the constraint $-2x_1 - 3x_2 \le 6$ will be perpendicular to $\nabla{f} = (-2, -3)$ and pass through the third quadrant with negative slope
* the boundary of the constraint $-2x_1 + 3x_2 \le 6$ will be perpendicular to $\nabla{f} = (-2, 3)$ and pass through the second quadrant with positive slope

Gradients also make it easy to check which side of an inequality is the feasible side (i.e., the side that satisfies the constraint.)
* the feasible side of the boundary of $f(x) \ge b$ is the side that $\nabla{f}$ points to
* the feasible side of the boundary of $f(x) \le b$ is the opposite side from what $\nabla{f}$ points to (i.e., the side $-\nabla{f}$ points to)
* the feasible set of $f(x) = b$ is just the line itself

1.15 - graphing feasible sets



---

## Terms

* Affine Function
* Constraint
* Gradient
* Linear Constraint
* Linear Function
* [LP] Linear Program
* Objective Function
* Mathematical Model

---